cowbird/services/impl/magpie.py

```diff
@@ -46,19 +114,56 @@ def permission_created(self, permission):
     def permission_deleted(self, permission):
         self.permissions_synch.delete_permission(permission)
 
-    def create_permission(self, permission):
-        # type: (Permission) -> None
+    def create_permission(self, permissions_data):
+        # type: (List[Dict[str,str]]) -> None
         """
         Make sure that the specified permission exists on Magpie.
-
-        .. todo:: First need to check if the permission already exists
-                  If the permission doesn't exist do a POST to create it
-                  If the permission exists but is different do a PUT to update it
         """
+        cookies = self.login()
 
-    def delete_permission(self, permission):
-        # type: (Permission) -> None
+        if permissions_data:
+            permissions_data[-1]["action"] = "create"
+
+            resp = requests.patch(
+                url=f"{self.url}/permissions",
+                headers=self.headers, cookies=cookies, json={"permissions": permissions_data}
+            )
+            if resp.status_code == 200:
+                LOGGER.info("Permission creation was successful.")
+            else:
+                raise HTTPError(f"Failed to create permissions : {resp.text}")
+        else:
+            LOGGER.warning("Empty permission data, no permissions to create.")
+
+    def delete_permission(self, permissions_data):
+        # type: (List[Dict[str,str]]) -> None
         """
         Remove the specified permission from Magpie if it exists.
         """
-        # TODO: Post a DELETE request, handle error silently if the permission doesn't exist
+        cookies = self.login()
+        if permissions_data:
+            permissions_data[-1]["action"] = "remove"
+
+            resp = requests.patch(
+                url=f"{self.url}/permissions",
+                headers=self.headers, cookies=cookies, json={"permissions": permissions_data}
+            )
+            if resp.status_code == 200:
+                LOGGER.info("Permission removal was successful.")
+            else:
+                raise HTTPError(f"Failed to remove permissions : {resp.text}")
+        else:
+            LOGGER.warning("Empty permission data, no permissions to remove.")
+
+    def login(self):
```

Comment:
Good idea!
Is caching here as simple as just saving the cookies and reusing them, or can the cookies become invalid over time?

I suppose I could always wrap the requests to Magpie to first try with the input cookies, and if the requests fails, reapply a login to get new cookies and retry the request. I could probably catch a specific HTTP error triggered by invalid cookies here to detect this case.

Reply:
You would need to consider some kind of "timeout" as they will expire at some point. 
The thing though is that cookies detain an "expire time" where, if you perform another request with them while they are "still technically valid" for some time, they will auto-refresh and you will get that refreshed value in the response.

If you feel fancy, you could store updated cookie values received after performing any request toward Magpie, and consider them invalid only if a long time passed since the last request. Otherwise, just keeping them for eg 60s since the last login until you force a re-login (regardless of if they were about to expire or not) could work also. I let you decide. Any delay would be better either way than login each time for many consecutive requests.

Reply:
Added timeout of 60s for cookies